File: arhiv-ui3/src/workspace/dto.ts
import { JSONObj, Obj } from '../scripts/utils';

export type WorkspaceRequest =
  | {
      typeName: 'ListDocuments';
      query: string;
      page: number;
    }
  | {
      typeName: 'GetStatus';
    }
  | {
      typeName: 'GetDocument';
      id: string;
    }
  | {
      typeName: 'ParseMarkup';
      markup: string;
    }
  | {
      typeName: 'SaveDocument';
      id: string;
      subtype: string;
      data: DocumentData;
    }
  | {
      typeName: 'CreateDocument';
      documentType: string;
      subtype: string;
      data: DocumentData;
    }
  | {
      typeName: 'EraseDocument';
      id: string;
    }
  | {
      typeName: 'ListDir';
      dir?: string;
      showHidden: boolean;
    }
  | {
      typeName: 'CreateAttachment';
      filePath: string;
    }
  | {
      typeName: 'Scrape';
      url: string;
    };

export type WorkspaceResponse =
  | {
      typeName: 'ListDocuments';
      documents: ListDocumentsResult[];
      hasMore: boolean;
    }
  | {
      typeName: 'GetStatus';
      status: string;
    }
  | {
      typeName: 'GetDocument';
      id: string;
      documentType: string;
      title: string;
      subtype: string;
      updatedAt: string;
      data: DocumentData;
      backrefs: DocumentBackref[];
    }
  | {
      typeName: 'ParseMarkup';
      ast: MarkupElement;
    }
  | {
      typeName: 'SaveDocument';
      errors?: SaveDocumentErrors;
    }
  | {
      typeName: 'CreateDocument';
      id?: string;
      errors?: SaveDocumentErrors;
    }
  | {
      typeName: 'EraseDocument';
    }
  | {
      typeName: 'ListDir';
      dir: string;
      entries: DirEntry[];
    }
  | {
      typeName: 'CreateAttachment';
      id: string;
    }
  | {
      typeName: 'Scrape';
      documents: ListDocumentsResult[];
    };

export type ListDocumentsResult = {
  id: string;
  documentType: string;
  subtype: string;
  title: string;
  updatedAt: string;
};

export type DocumentData = JSONObj;

export type DocumentBackref = {
  id: string;
  documentType: string;
  subtype: string;
  title: string;
};

export type DocumentFieldErrors = Obj<string[]>;

export type SaveDocumentErrors = {
  documentErrors: string[];
  fieldErrors: DocumentFieldErrors;
};

export type DirEntry =
  | {
      typeName: 'Dir';
      name: string;
      path: string;
      isReadable: boolean;
    }
  | {
      typeName: 'File';
      name: string;
      path: string;
      isReadable: boolean;
      size: number;
    }
  | {
      typeName: 'Symlink';
      name: string;
      path: string;
      isReadable: boolean;
      linksTo: string;
      size?: number;
    };

export type MarkupElement =
  | {
      typeName: 'Document';
      children: MarkupElement[];
    }
  | {
      typeName: 'Text';
      range: Range;
      value: string;
    }
  | {
      typeName: 'Code';
      range: Range;
      value: string;
    }
  | {
      typeName: 'Html';
      range: Range;
      value: string;
    }
  | {
      typeName: 'FootnoteReference';
      range: Range;
      label: string;
    }
  | {
      typeName: 'SoftBreak';
      range: Range;
    }
  | {
      typeName: 'HardBreak';
      range: Range;
    }
  | {
      typeName: 'Rule';
      range: Range;
    }
  | {
      typeName: 'TaskListMarker';
      range: Range;
      checked: boolean;
    }
  | {
      typeName: 'Paragraph';
      range: Range;
      children: MarkupElement[];
    }
  | {
      typeName: 'Heading';
      range: Range;
      level: 'H1' | 'H2' | 'H3' | 'H4' | 'H5' | 'H6';
      children: MarkupElement[];
    }
  | {
      typeName: 'BlockQuote';
      range: Range;
      children: MarkupElement[];
    }
  | {
      typeName: 'CodeBlock';
      range: Range;
      kind: 'Indented' | { Fenced: string };
      children: MarkupElement[];
    }
  | {
      typeName: 'List';
      range: Range;
      first_item_number?: number;
      children: MarkupElement[];
    }
  | {
      typeName: 'ListItem';
      range: Range;
      children: MarkupElement[];
    }
  | {
      typeName: 'FootnoteDefinition';
      range: Range;
      label: string;
      children: MarkupElement[];
    }
  | {
      typeName: 'Table';
      range: Range;
      alignments: Array<'None' | 'Left' | 'Center' | 'Right'>;
      children: MarkupElement[];
    }
  | {
      typeName: 'TableHead';
      range: Range;
      children: MarkupElement[];
    }
  | {
      typeName: 'TableRow';
      range: Range;
      children: MarkupElement[];
    }
  | {
      typeName: 'TableCell';
      range: Range;
      children: MarkupElement[];
    }
  | {
      typeName: 'Emphasis';
      range: Range;
      children: MarkupElement[];
    }
  | {
      typeName: 'Strong';
      range: Range;
      children: MarkupElement[];
    }
  | {
      typeName: 'Strikethrough';
      range: Range;
      children: MarkupElement[];
    }
  | {
      typeName: 'Link';
      range: Range;
      link_type: LinkType;
      url: string;
      title: string;
      children: MarkupElement[];
    }
  | {
      typeName: 'Image';
      range: Range;
      link_type: LinkType;
      url: string;
      title: string;
      children: MarkupElement[];
    };

export type Range = {
  start: number;
  end: number;
};

type LinkType =
  | 'Inline'
  | 'Reference'
  | 'ReferenceUnknown'
  | 'Collapsed'
  | 'CollapsedUnknown'
  | 'Shortcut'
  | 'ShortcutUnknown'
  | 'Autolink'
  | 'Email';

export function throwBadMarkupElement(value: never): never;
export function throwBadMarkupElement(value: MarkupElement) {
  throw new Error(`Unknown MarkupElement: ${value.typeName}`);
}

File: arhiv-ui3/src/workspace/dto.rs
use std::collections::HashMap;

use serde::{Deserialize, Serialize};

use arhiv_core::entities::{DocumentData, Id};
use rs_utils::Timestamp;

#[derive(Deserialize, Debug)]
#[serde(deny_unknown_fields, tag = "typeName")]
pub enum WorkspaceRequest {
    ListDocuments {
        query: String,
        page: u8,
    },
    GetStatus {},
    GetDocument {
        id: Id,
    },
    RenderMarkup {
        markup: String,
    },
    #[serde(rename_all = "camelCase")]
    CreateDocument {
        document_type: String,
        subtype: String,
        data: DocumentData,
    },
    SaveDocument {
        id: Id,
        subtype: String,
        data: DocumentData,
    },
    EraseDocument {
        id: Id,
    },
    #[serde(rename_all = "camelCase")]
    ListDir {
        dir: Option<String>,
        show_hidden: bool,
    },
    #[serde(rename_all = "camelCase")]
    CreateAttachment {
        file_path: String,
    },
    Scrape {
        url: String,
    },
}

#[derive(Serialize)]
#[serde(deny_unknown_fields, tag = "typeName")]
pub enum WorkspaceResponse {
    ListDocuments {
        documents: Vec<ListDocumentsResult>,
        #[serde(rename = "hasMore")]
        has_more: bool,
    },
    GetStatus {
        status: String,
    },
    #[serde(rename_all = "camelCase")]
    GetDocument {
        id: Id,
        title: String,
        document_type: String,
        subtype: String,
        updated_at: Timestamp,
        data: DocumentData,
        backrefs: Vec<DocumentBackref>,
    },
    RenderMarkup {
        html: String,
    },
    CreateDocument {
        id: Option<Id>,
        errors: Option<SaveDocumentErrors>,
    },
    SaveDocument {
        errors: Option<SaveDocumentErrors>,
    },
    EraseDocument {},
    ListDir {
        dir: String,
        entries: Vec<DirEntry>,
    },
    CreateAttachment {
        id: Id,
    },
    Scrape {
        documents: Vec<ListDocumentsResult>,
    },
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ListDocumentsResult {
    pub id: Id,
    pub document_type: String,
    pub subtype: String,
    pub title: String,
    pub updated_at: Timestamp,
}

#[derive(Serialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct SaveDocumentErrors {
    pub document_errors: Vec<String>,
    pub field_errors: HashMap<String, Vec<String>>,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DocumentBackref {
    pub id: Id,
    pub document_type: String,
    pub subtype: String,
    pub title: String,
}

#[derive(Serialize)]
#[serde(tag = "typeName")]
pub enum DirEntry {
    #[serde(rename_all = "camelCase")]
    Dir {
        name: String,
        path: String,
        is_readable: bool,
    },

    #[serde(rename_all = "camelCase")]
    File {
        name: String,
        path: String,
        is_readable: bool,
        size: u64,
    },

    #[serde(rename_all = "camelCase")]
    Symlink {
        name: String,
        path: String,
        is_readable: bool,
        links_to: String,
        size: Option<u64>,
    },
}

impl DirEntry {
    pub fn get_name(&self) -> &str {
        match self {
            DirEntry::Dir { name, .. } => name,
            DirEntry::File { name, .. } => name,
            DirEntry::Symlink { name, .. } => name,
        }
    }
}
